### ISSUE TITLE
Attach interactively to docker container

### ISSUE THREAD

**Issue by adaroobi** [mintty/wsltty#41]
As the title, Could not `attach` or `exec` using `--interactive` STDIN with error
the input device is not a TTY

**Comment by mintty** [comment]
Not the slightest idea what your actual use case is and what you want.
How should wsltty be related to Docker?

**Comment by adaroobi** [comment]
Well, maybe an example could help.
Executing `docker.exe run -it ubuntu bash`

Expected: Enter the container that's created and interact with with its prompt and keep the current stdin open

Actual: 
> the input device is not a TTY.  If you are using mintty, try prefixing the command with 'winpty'

It can be executed only when omitting the -i flag but, this will no more keep the stdin open which reduces some usability.

**Comment by mintty** [comment]
To start wsltty, you would have to somehow invoke wsltty, not just ubuntu bash.
Also, since mintty is a stand-alone Windows program, it does not use any stdin.
Where's the quoted output coming from, what would -i mean here... Your intent is obscure.
Unless you describe what you want clearly, and without assuming any knowledge about docker, I can only close the issue.

**Comment by adaroobi** [comment]
I've started wsltty the way its described in the readme so its not a case of misuse but can be misunderstanding, an issue or something else, however I'll close this myself and reopen it when again when I am able to get an isolated example may help us both determine the intent.

Thanks.

**Comment by mamiu** [comment]
@mintty  
  
Imagine following Situation:
* Windows as development machine
* Docker installed on Windows
* Using docker through WSL
* wsltty as terminal emulator

If I want to log into a docker container (which is a usual task during the use of docker. have a look here: [https://askubuntu.com/a/507009](https://askubuntu.com/a/507009)) I'd type this command: **docker.exe exec -i -t <container> /bin/bash**.

But then, I'll get following error:
```the input device is not a TTY.  If you are using mintty, try prefixing the command with 'winpty'```

After prefixing 'winpty' the 'command not found' message appears.
```winpty: command not found```

**Comment by mintty** [comment]
> winpty: command not found

So I suggest to install https://github.com/rprichard/winpty.
I agree it would be a good idea to integrate its functionality automatically, but that's a cygwin issue, not a mintty issue.
Actually, we may be having an additional problem layer here; I assume you invoke docker from WSL, within wsltty? In that case, actually, winpty should already be involved, in its incarnation as wslbridge. Don't know how to interpret the issue then; please clarify your workchain.

**Comment by mintty** [comment]
There's also https://github.com/tiangolo/babun-docker
which may be helpful, provided you try to run docker from cygwin, which isn't clear from your description.

**Comment by adaroobi** [comment]
@mintty I tried both suggestions before, but no successful result. I didn't get a time to dig further
and used cmd/ps instead...
Seems that I am not the only one who is experiencing the same symptoms.

**Comment by mamiu** [comment]
I don't use cygwin. I just installed WSL and instead using the old fashioned cmd for the ubuntu bash I also installed wsltty, which btw is awesome!
I used it for a while and everything worked fine. Than, somebody came up with docker for windows. So I also installed docker for windows to improve my development workflow, but docker doesn't run smoothly within wsltty. First I had some codepage issues, which I resolved with ```chcp.com 932```. Everything else was running frictionless again. But logging into a docker container is the only thing, that doesn't work so far.

I think more cannot be said. If you still have questions, please ask! :)

**Comment by mintty** [comment]
Your codepage issue is probably an instance of mintty/mintty#376 and the I/O interworking one of mintty/mintty#56, both not really being mintty issues.

So you invoke Windows docker from WSL (inside wsltty), right?
I wasn't aware this is possible at all.
This seems to be the reverse interworking case of what wslbridge accomplishes within wsltty;
@rprichard, any idea about the issue?

**Comment by rprichard** [comment]
FWIW: wslbridge and winpty aren't that related, though I wrote both.  wslbridge is more robust.

 * wslbridge: bridges between two Unix ptys

 * winpty: connects a Windows console to a Cygwin/MSYS pty

WSL can run console programs, but in my limited testing, it didn't work very well.  A WSL pty was regarded as a `FILE_TYPE_CHAR` device, so `isatty` returned true for C stdio handles.  Ctrl-C didn't work.  I tried a few console-specific APIs, and they didn't work with the WSL pty handles (understandably).  A WSL pty only generates VT/100 input, so control/navigation keys didn't work with the programs I tried (cmd, powershell, python3).

I looked into Docker/winpty integration a while ago.  IIRC, Docker had a client program that didn't use Cygwin/MSYS, and it worked something like a Windows SSH client -- if its stdio were connected to a console, then it'd create a pty in the Linux VM and handle escape sequences, and otherwise, it would just copy bytes between the Linux VM and the Windows STDIN/STDOUT/STDERR file handles.  If you tried to run Docker in mintty with a Cygwin pty, then it'd kinda work, except that it wouldn't create a Linux pty?

wslbridge has a Cygwin/MSYS program frontend that connects to a WSL pty, but it wouldn't be that hard to flip it around.  You could use inverted-wslbridge to connect from wsltty to a Cygwin pty, then run winpty inside that.  That's 4 adapters to the Linux VM, though:

 * `mintty.exe` (Cygwin pty)
    * `wslbridge.exe` (Cygwin pty -> WSL pty)
       * `inverted-wslbridge.exe` (WSL pty -> Cygwin pty)
          * `winpty.exe` (Cygwin pty -> console)
             * `docker.exe`? (console -> Linux pty)

Maybe it'd be OK -- wslbridge mostly just adds latency.  `winpty.exe` and `docker.exe` would cause most of the trouble.  `inverted-wslbridge` would be a WSL program that runs a Cygwin/MSYS `inverted-wslbridge-backend.exe` program, and you'd also need a `winpty.exe` built for the same Cygwin/MSYS environment.

A WSL winpty adapter might be possible.  It would be easier if WSL could access Win32 named pipes.  (Maybe it can, somehow?)  Maybe winpty could use sockets instead of named pipes.

It could also make sense to have a WSL docker adapter (i.e. WSL pty -> Linux pty).  I'm wondering what's happening in the Docker VM when `docker.exe` connects.

**Comment by raupie** [comment]
I have the same exact issue. 

Installed Docker for Windows. Enabled WSL. Exposed the TCP port. Created windows Firewall rule. 

Can run docker.exe commands however `exec` fails with `the input device is not a TTY. If you are using mintty, try prefixing the command with winpty`.

I installed wsltty, ran it from the context menu, still getting the same message. Am I missing something?

**Comment by mintty** [comment]
This issue is related to subtle mutual interaction between pty-based and console-based programs. See rprichard's comment above for some technical details.
It is not quite clear from your description what exactly fails (`exec` where) and what/how it does not fail.
However, that wouldn't help as long as there is no `inverted-wslbridge`. Also note that this is not a mintty-specific issue, you might cross-check with xterm.

**Comment by mintty** [comment]
Does docker have an interactive mode? (A kind of shell?)
In that case, mintty could perhaps run docker directly, probably with winpty. Not in the same terminal as WSL, though.
Maybe WSLtty could bundle winpty anyway, so you could also launch e.g. a cmd.exe session in parallel.
@rprichard, any comments?

**Comment by rprichard** [comment]
It could make sense to package winpty with wsltty, I suppose. You'd need a special shortcut that launches mintty with cmd using winpty. Needing to start a new terminal window to connect to a Docker instance strikes me as disruptive to one's workflow, but maybe it's OK?

It's unfortunate that the Docker instance and mintty both use VT100 I/O, but we have to bridge between the two with console I/O, a translation process that's lossy and adds latency.

**Comment by mintty** [comment]
Not really belonging here, but: I noticed again, on this occasion, that with winpty cmd.exe ^C does not work (which is better than quitting cmd.exe, of course). Any chance to fix that?

**Comment by rprichard** [comment]
winpty's Ctrl-C problem is tracked here, https://github.com/rprichard/winpty/issues/116. It's hard to fix.

**Comment by kevin-wang-cy** [comment]
I got same issue when I try docker -it command in the bash on Ubuntu on Windows (WLS) , it report below errors:
{quote}
the input device is not a TTY.  If you are using mintty, try prefixing the command with 'winpty
{quote}

Hope it can be fixed soon

**Comment by mintty** [comment]
What {quote}?
And the technical conditions for a "fix" have been discussed here; there is currently no fix in sight. It's a multiple-interworking issue.

**Comment by caot** [comment]
Here is how to make it work:
1. build https://github.com/rprichard/winpty
2. winpty docker run -ti --rm <image REPOSITORY name>:latest sh

**Comment by whereisaaron** [comment]
I had this problem with the original WSL on Windows. I couldn't work out how to get rid of `mintty` and `winpty` was not installed, and not desirable anyway. Installing the new from Windows store version fixed the probably and the shell worked as a tty. I had to reinstall, but this is much better.

```
PS C:\> wslconfig.exe /list
Windows Subsystem for Linux Distributions:
Legacy (Default)  <---- tty problem
Ubuntu  <---- works! :)
```

**Comment by whereisaaron** [comment]
Gah, I copied my user files across and now I have the same problem with the new Windows Store Ubuntu. So could it be a `shopt` setting causing this? This is on WSL with 1803.

Confirmed running `docker exec -it` from a plain/new bash login works, with my `.bashrc` in place it fails, but if I run it with `sh` (which doesn't load the `.bashrc`) is works again. So it a actually a bash setting combined with WSL (as these bash setting work on my Linux hosts).

![image](https://user-images.githubusercontent.com/1109585/46779773-3eec8600-cce7-11e8-9b6f-dcd0a888be13.png)
